os/drivers/seclink/seclink_drv_key.c

```diff
@@ -29,71 +29,80 @@
 #endif
 #define SLDRV_TAG "[SECLINK_DRV_KEY]"
 
-#define NKEY_SLOT 4
+#define NKEY_SLOT 64
+/*
+ * +-----------------+------------------+
+ * |    ROM (0~31)   | RAM(32~63)       |
+ * +-----------------+------------------+
+ */
+#define KEY_ROM_ENTRY 0
+#define KEY_RAM_ENTRY 32
 static uint8_t g_key_inuse[NKEY_SLOT] = {0,};
+#define KEYSLOT_USE 0
+#define KEYSLOT_NOTUSE 1
 
-static int _set_key_slot(uint8_t num)
+static inline int _is_rom_area(uint8_t num)
 {
-	if (num >= NKEY_SLOT) {
-		return -1;
+	if (num < KEY_RAM_ENTRY) {
+		return 1;
 	}
-	g_key_inuse[num] = 1;
 	return 0;
 }
 
 
-static int _unset_key_slot(uint8_t num)
+static inline int _set_key_slot(uint8_t num)
 {
 	if (num >= NKEY_SLOT) {
 		return -1;
 	}
-	g_key_inuse[num] = 0;
+	g_key_inuse[num] = KEYSLOT_USE;
 	return 0;
 }
 
 
-static int _is_empty_key_slot(uint8_t num)
+static inline int _unset_key_slot(uint8_t num)
 {
 	if (num >= NKEY_SLOT) {
 		return -1;
 	}
-
-	return g_key_inuse[num] == 0 ? 1 : 0;
+	g_key_inuse[num] = KEYSLOT_NOTUSE;
+	return 0;
 }
 
 
-void hd_initialize_key_slot(void)
+static inline int _is_empty_key_slot(uint8_t num)
 {
-	for (int i = 0; i < NKEY_SLOT; i++) {
-		g_key_inuse[i] = 0;
+	if (num >= NKEY_SLOT) {
+		return -1;
 	}
+
+	return g_key_inuse[num];
 }
 
 
-int hd_handle_key_request(int cmd, unsigned long arg, void *lower)
+static int hd_handle_rom_key(int cmd, struct seclink_req *req, struct seclink_key_info *info, struct sec_lowerhalf_s *se)
 {
-	SLDRV_ENTER;
-
-	struct seclink_req *req = (struct seclink_req *)arg;
-	if (!req) {
-		return -EINVAL;
-	}
-
-	struct seclink_key_info *info = (struct seclink_key_info *)req->req_type.key;
-	if (!info) {
-		return -EINVAL;
+	if (cmd != SECLINKIOC_GETKEY) {
+		req->res = HAL_INVALID_REQUEST;
+		return 0;
 	}
+	req->res = se->ops->get_key(info->mode, info->key_idx, info->key);
```

Comment:
1. it should check validation of req & se, & info.
2. Below is needed, because all lower layer doesn't implement all of lower's ops, like a abstract class. 
```
if (se->ops->get_key == NULL) {
     return -ENOSYS;
}
```